### PR TITLE
sela- started working on general opmode + added servo to turn scoring…

### PR DESCRIPTION
**Pull request by boozmcooz7**
… unit
אין לי מושג מה עשיתי בטלאופ,חירטטתי פקודות אז שהקוד יפסיק לצרוח עליי
Before issuing a pull request, please see the contributing page.

### CODE REVIEW

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java [hunk 1]
```diff
@@ -5,30 +5,66 @@
...
 
-    // intake down and open the intake claw
     public static Command IntakeCommand() {
-        return new ParallelCommandGroup(
+        return new SequentialCommandGroup(mmSystems.linearIntake.setPosition(linearIntakeOpen),
+                new WaitCommand(timeLinearIntake),
+                new ParallelCommandGroup(
                 mmSystems.intakeArm.intakeDown(),
-                mmSystems.intakEndUnit.openIntakeClaw());
+                mmSystems.intakEndUnit.openIntakeClaw()), new WaitCommand(2*timeLinearIntake));
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java]: what is the use of `new WaitCommand(2*timeLinearIntake)`?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java [hunk 2]
```diff
@@ -5,30 +5,66 @@
...
+    /* intake done  -
+    1. close claw
+    2. intake arm up  ,
+    3. close linear intake */
     public static Command IntakeDoneCommand() {
         return new SequentialCommandGroup(
                 mmSystems.intakEndUnit.closeIntakeClaw(),
-                mmSystems.intakeArm.intakeUp());
+                mmSystems.intakeArm.intakeUp()).
+                andThen(mmSystems.linearIntake.setPosition(0));
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java]: replace 0 with a constant

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java [hunk 3]
```diff
@@ -5,30 +5,66 @@
...
+    /* score sample-
+    1. elevator get to desired height
+    2. scoring servo turn
+    3. scoring claw open
+     */
+    public static Command ScoreSample(){
+        return new SequentialCommandGroup(
+                mmSystems.elevator.moveToPose(Elevator.HIGH_BASKET),
+                mmSystems.scoringEndUnit.scoreScoringServo(),
+                mmSystems.scoringEndUnit.openScoringClaw(),new WaitCommand(timeLinearIntake));
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java]: why not do it in parallel?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleOp.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+public class TeleOp extends MMOpMode {
+    private GamepadEx gamepadEx;
+    public TeleOp(OpModeType.NonCompetition opModeType) {
+        super(opModeType);
+    }
+
+    @Override
+    public void onInit() {
+        MMRobot.getInstance().mmSystems.initRobotSystems();
+        gamepadEx = new GamepadEx(gamepad1);
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleOp.java]: no need to create a new gamepad take the ones from MMSystems

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleOp.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            RobotCommands.ScoreSample();
+            RobotCommands.HoldPosScoring();
+        }
+
+
+        // button y -total intake
+        if (gamepadEx.wasJustPressed(GamepadKeys.Button.Y)){
+            RobotCommands.IntakeCommand();
+            RobotCommands.IntakeDoneCommand();
+        }
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleOp.java]: create triggers

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleOp.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        waitForStart();
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleOp.java]: remove this(my bad)

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java [hunk 7]
```diff
@@ -10,21 +10,30 @@
 
 public class ScoringEndUnit extends SubsystemBase {
     CuttleServo clawscoringServo;
+    CuttleServo posscoringServo;
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java]: `posscoringServo` did you mean `scoringServo`?

**Review by ori-coval — CHANGES_REQUESTED** [review]

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java (hunk 1 above)
Comment (boozmcooz7) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java]: after the claw is opened i want the command to wait (until sample is absorbed), I don't have a way of ensuring that yet so I put a wait to command

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java (hunk 7 above)
Comment (boozmcooz7) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java]: i wanted it to be clear it was the position servo and not the claw

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java [hunk 8]
```diff
@@ -5,30 +5,67 @@
...
 
 import java.util.Collections;
 import java.util.Set;
 
 public class RobotCommands {
 
     private static MMSystems mmSystems = MMRobot.getInstance().mmSystems;
+    private static int timeLinearIntake= 500; //millis
+    private static double linearIntakeOpen = 2.22;
+    private  static double linearIntakeClosed = 0;
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java]: make the constants final

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java [hunk 9]
```diff
@@ -1,34 +1,71 @@
...
-import java.util.Collections;
-import java.util.Set;
+import java.util.function.DoubleSupplier;
 
 public class RobotCommands {
 
-    private static MMSystems mmSystems = MMRobot.getInstance().mmSystems;
+    private static final MMSystems mmSystems = MMRobot.getInstance().mmSystems;
+    private static final int timeLinearIntake = 500; //millis
+    private static final double linearIntakeOpen = 2.22;
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java]: linearIntakeOpen  isnt used remove it

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java [hunk 10]
```diff
@@ -1,34 +1,71 @@
...
+    private static final double linearIntakeClosed = 0;
+    private final double maxOpening = 0.23;
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java]: maxOpening isnt used remove it

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java [hunk 11]
```diff
@@ -1,34 +1,71 @@
...
+    /* intake recieve -
+    1. open linear intake
+    2. wait  ,
+    3. intake down and open claw*/
+
+    public static Command IntakeCommand(DoubleSupplier intake_trigger) {
+        return new SequentialCommandGroup(mmSystems.linearIntake.setPositionByJoystick(joysxlj),
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java]: `joysxlj`?!

also the command need to happen in parallel so the intake end unit would go down and open while the linear intake opens

also currently this wouldnt even work Sequentialy cuse the setPositionByJoystick never ends cuse its a run command

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java [hunk 12]
```diff
@@ -1,34 +1,71 @@
...
+                new WaitCommand(timeLinearIntake),
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java]: because the linearIntake open to a location by the trigger there is no need to wait for it to open so remove this

also like in my erlier comment the this wont work because the setPositionByJoystick neev

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java [hunk 13]
```diff
@@ -1,34 +1,71 @@
...
+    /* intake done  -
+    1. close claw
+    2. intake arm up  ,
+    3. close linear intake */
     public static Command IntakeDoneCommand() {
         return new SequentialCommandGroup(
                 mmSystems.intakEndUnit.closeIntakeClaw(),
-                mmSystems.intakeArm.intakeUp());
+                mmSystems.intakeArm.intakeUp()).
+                andThen(mmSystems.linearIntake.setPosition(linearIntakeClosed));
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CommandGroup/RobotCommands.java]: you should wait a bit for the endunit to close before moving it up

also why not have intakeUp and move the linearIntake at the same time it shouldnt be a problem

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java (hunk 7 above)
Comment (ori-coval) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java]: i think scoringArmServo would be a better name

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java [hunk 14]
```diff
@@ -10,12 +10,16 @@
...
 
     public static double open = 1;
     public static double close = -1;
+    public static double scoring_hold = -1;
+    public static double scoring_score = 1;
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java]: not clear for what servo this is i would recommand adding arm to the name of this constant

Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java]: constants should be final

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java [hunk 15]
```diff
@@ -10,12 +10,16 @@
 
 public class ScoringEndUnit extends SubsystemBase {
     CuttleServo clawscoringServo;
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java:12]: `clawscoringServo` upper case S in scoring

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java [hunk 16]
```diff
@@ -27,4 +31,12 @@ public Command closeScoringClaw() {
...
 
     }
+
+    public Command scoreScoringServo() {
+        return new InstantCommand(() -> posscoringServo.setPosition(scoring_score), this);
+    }
+
+    public Command holdScoringServo() {
+        return new InstantCommand(() -> posscoringServo.setPosition(scoring_hold), this);
+    }
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/SubSystems/ScoringEndUnit.java]: both functions names are unclear change to something more like

scoringArmScore

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+import com.arcrobotics.ftclib.gamepad.GamepadKeys;
+
+import org.firstinspires.ftc.teamcode.CommandGroup.RobotCommands;
+import org.firstinspires.ftc.teamcode.Libraries.MMLib.MMOpMode;
+import org.firstinspires.ftc.teamcode.MMRobot;
+import org.firstinspires.ftc.teamcode.MMSystems;
+import org.firstinspires.ftc.teamcode.SubSystems.ScoringEndUnit;
+import org.firstinspires.ftc.teamcode.utils.OpModeType;
+
+public class TeleopGeneral extends MMOpMode {
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java]: `@TeleOp` is missing

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+
+    public TeleopGeneral(OpModeType.NonCompetition opModeType) {
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java]: in `public TeleopGeneral(OpModeType.NonCompetition opModeType) {` no need for `OpModeType.NonCompetition opModeType` remove it 

to be clear only remove it in the builder keep the one in the super

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+        super(OpModeType.NonCompetition.EXPERIMENTING_NO_EXPANSION);
+    }
+
+    @Override
+    public void onInit() {
+        MMRobot.getInstance().mmSystems.initRobotSystems();
+
+        Trigger prepareScore = new Trigger(
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java]: prepare score to where high basket? low basket? specimen?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java:1]: on all the triggers add `Trigger` at the end so it would be clear its a trigger

for example: `intakeOpenTrigger`

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+                () -> MMRobot.getInstance().mmSystems.gamepadEx1.getButton(GamepadKeys.Button.X)
+        );
+
+        Trigger scoreSample = new Trigger(
+                () -> MMRobot.getInstance().mmSystems.gamepadEx1.getButton(GamepadKeys.Button.Y)
+        );
+
+        Trigger totalIntake = new Trigger(
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java]: why did you add the total? it makes it unclear what the trigger is for

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+        prepareScore.whileActiveOnce(RobotCommands.PrepareHighSample());
+
+        //new button that score- y
+        scoreSample.whileActiveOnce(MMRobot.getInstance().mmSystems.scoringEndUnit.openScoringClaw());
+
+
+        // intake using trigger
+        totalIntake.whileActiveOnce(
+                RobotCommands.IntakeCommand(() -> MMRobot.getInstance().mmSystems.gamepadEx1.getTrigger(GamepadKeys.Trigger.RIGHT_TRIGGER)).whenFinished(
+                        () -> RobotCommands.IntakeDoneCommand().schedule()));
```
Comment (ori-coval) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TeleOp/TeleopGeneral.java]: you need to have a meeting of Kanna to choose all the robot function and buttons the driver would have

this need to be done soon so you can know how to create the buttons and commands

**Review by ori-coval — CHANGES_REQUESTED** [review]
